tracing/span.go

```diff
@@ -209,91 +162,162 @@ func (s *Span) ToZipkinSpan() ZipkinSpan {
 	return zs
 }
 
-func (s *Span) isDebugSet() bool {
-	return s.flags&FlagMaskDebug != 0
+func (t *trace) isDebugSet() bool {
+	return t.flags&FlagMaskDebug != 0
 }
 
-// ShouldSample returns true if this span should be sampled.
+func (t *trace) setDebug(v bool) {
+	if v {
+		t.flags |= FlagMaskDebug
+	} else {
+		t.flags &= ^FlagMaskDebug
+	}
+}
+
+// shouldSample returns true if this span should be sampled.
 //
 // If the span's debug flag was set, then this function will always return true.
-func (s *Span) ShouldSample() bool {
-	return s.sampled || s.isDebugSet()
+func (t *trace) shouldSample() bool {
+	return t.sampled || t.isDebugSet()
 }
 
-// End ends a span.
-//
-// If err is non-nil, the error tag will also be set to true.
-//
-// It also sends the span into tracer, if the span should be sampled.
-// The context object passed in is used to control the timeout of the sending,
-// see Tracer.MaxRecordTimeout for more info.
-func (s *Span) End(ctx context.Context, err error) error {
-	s.end = time.Now()
+func (t *trace) stopAndPublish(ctx context.Context, err error) error {
+	t.stop = time.Now()
 
-	if err != nil {
-		s.SetTag(ZipkinBinaryAnnotationKeyError, true)
-		if s.spanType == SpanTypeServer && errors.Is(err, context.DeadlineExceeded) {
-			s.SetTag(ZipkinBinaryAnnotationKeyTimeOut, true)
-		}
+	if !t.shouldSample() || t.tracer == nil {
+		return nil
 	}
-	if s.isDebugSet() {
-		s.SetTag(ZipkinBinaryAnnotationKeyDebug, true)
+
+	if err != nil {
+		t.setTag(ZipkinBinaryAnnotationKeyError, true)
 	}
-	if s.spanType == SpanTypeLocal && s.component != "" {
-		s.SetTag(ZipkinBinaryAnnotationKeyLocalComponent, s.component)
+
+	if t.isDebugSet() {
+		t.setTag(ZipkinBinaryAnnotationKeyDebug, true)
 	}
-	for _, hook := range s.hooks {
-		if hookErr := hook.OnEnd(s, err); hookErr != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnEnd hook error: " + hookErr.Error())
+
+	return t.tracer.Record(ctx, t.toZipkinSpan())
+}
+
+// Span interface
+type Span interface {
+	Name() string
+	SpanType() SpanType
+	Hooks() []SpanHook
+	AddHooks(hooks ...SpanHook)
+	Start()
+	Stop(err error)
+	Publish(ctx context.Context, err error) error
+	SetTag(key string, value interface{})
+	AddCounter(key string, delta float64)
+	CreateLocalChild(name, component string) *LocalSpan
+	CreateClientChild(name string) *ClientSpan
+	LogError(msg string, err error)
+	SetDebug(v bool)
+}
+
+// StartSpan starts the Span and calls all OnStart Hooks registered to the Span.
+func StartSpan(s Span) {
```

Comment:
Is there a reason we need to use this function over `s.Start()`? Same as other similar functions below.

Reply:
These functions are what actually call the `Hooks` so you don't have to remember to do that if you want to do custom things in your implementation.

Reply:
That sounds error prone, as the user is very likely to accidentally call `s.Span` instead of `StartSpan(s)`.

The interface here is slightly different from hooks. For hooks we provide an interface for the users to do their own implementations. Here we provide an interface because we provided several implementations with the same interface. So here I think we can just make sure the type functions do all the things (including calling the hooks) and get rid of these global functions.

Reply:
Fair, I can change that.